### PR TITLE
Add dynamodb boolean datatype

### PR DESCRIPTION
**Pull request by jamesthompson**
This PR adds the following (following on from the issue described here: https://github.com/aristidb/aws/issues/173)

* Updated nix settings for build with the old errors package (superseded on the main stable nix-channel).
* Support for the native dynamodb boolean data type.

This work fixes the issue that the latest dynamodb json responses would cause an exception to be thrown when FromDynItem getAttr was called on the response set if boolean values are encoded.

I've tried to keep the code as close to the style for the other data types. It would benefit from some close scrutiny of the DynSize instance - which I'm a little unsure of.

### CODE REVIEW

File: shell.nix [hunk 1]
```diff
@@ -1,37 +1,5 @@
...
-                 aeson base bytestring errors http-client lifted-base monad-control
-                 mtl QuickCheck quickcheck-instances resourcet tagged tasty
-                 tasty-quickcheck text time transformers transformers-base
-               ];
-               homepage = "http://github.com/aristidb/aws";
-               description = "Amazon Web Services (AWS) for Haskell";
-               license = stdenv.lib.licenses.bsd3;
-             }) {};
+let
+  pkg = haskellngPackages.callPackage ./. {};
```
Comment (aristidb) [inline shell.nix]: Did peti change the output of cabal2nix --shell? oO

File: aws.cabal [hunk 2]
```diff
@@ -285,7 +285,7 @@ Executable Sqs
     Build-depends:
                        base == 4.*,
                        aws,
-                       errors >= 1.4 && < 2.0,
+                       errors >= 1.4.7 && <= 2.0,
```
Comment (aristidb) [inline aws.cabal]: That should be < 2.0.

File: aws.cabal [hunk 3]
```diff
@@ -1,5 +1,5 @@
 Name:                aws
-Version:             0.12.1
+Version:             0.12.2
```
Comment (aristidb) [inline aws.cabal]: Your change is API breaking, technically. Let's make it 0.13.

File: Aws/DynamoDb/Core.hs [hunk 4]
```diff
@@ -401,18 +420,6 @@ fromTS i = UTCTime (ModifiedJulianDay days) diff
...
--- | Encoded as 0 and 1.
-instance DynVal Bool where
-    type DynRep Bool = DynNumber
-    fromRep (DynNumber i) = do
-        (i' :: Int) <- toIntegral i
-        case i' of
-          0 -> return False
-          1 -> return True
-          _ -> Nothing
-    toRep b = DynNumber (if b then 1 else 0)
```
Comment (aristidb) [inline Aws/DynamoDb/Core.hs]: I'd like @ozataman to comment on the impact of losing this instance.

**Comment by aristidb** [comment]
I've added a few inline comments.

@ozataman any comments on this change?

File: shell.nix (hunk 1 above)
Comment (jamesthompson) [inline shell.nix]: I just split out the shell.nix file. I think this is pretty standard these days.. Can revert if it's preferred...

File: aws.cabal (hunk 3 above)
Comment (jamesthompson) [inline aws.cabal]: Will bump.

File: shell.nix (hunk 1 above)
Comment (aristidb) [inline shell.nix]: The file was automatically generated. I'd prefer to be able to automatically regenerate it in case of updates / changes. :) Do you know if there is a way to do that in this splitted form?

Comment (jamesthompson) [inline shell.nix]: All you have to do is regenerate the default.nix when you need to update it. In the case that the errors lib gets moved up to the present nix-channel version (2.0 I believe) you would just delete config.nix.

In order to regenerate the default.nix, you would run `cabal2nix . > default.nix`.

Comment (aristidb) [inline shell.nix]: No, sounds good actually. No need to revert. :)

**Comment by aristidb** [comment]
As @ozataman has seemingly not had a chance yet to look at this, maybe you can shed a light on my major worry with this, @jamesthompson:

There was already an instance of Bool that used numbers. Will the loss/change of this instance cause existing programs to break?

**Comment by jamesthompson** [comment]
@aristidb The instance was (I believe) a means to just deserialize boolean values encoded as 0 or 1. Native boolean json keyed with the type `"BOOL"` in the dynamo response json caused parsing exceptions. This code adds support for the now native dynamo boolean (i.e. json boolean) data which is returned from dynamo.

This change would require people to refactor their parsers accordingly (hence, I believe the major api bump is necessary). 

tl;dr: If your code maps the json numbers 0 or 1 to booleans written against the present API - you will have to make a change and manually roll a parser for those numbers, mapping them to booleans. Alternatively - you would change your dynamo schema to make use of the supported booleans.

**Comment by jamesthompson** [comment]
Hi @aristidb,  was wondering if you've heard anything about this PR? Do let me know if I can be of any assistance...

**Comment by aristidb** [comment]
I'm going to merge this now, as @ozataman has not weighed in after 20 days. If he strongly objects, I may still unmerge it though.

**Comment by ozataman** [comment]
Sorry for the delay folks, been hammered at work lately. Sounds good - will review this soon and reply here.

**Comment by ozataman** [comment]
Just looked at the Bool changes while in transit. Thanks for contributing that bit - totally agree that we should keep up to date with aws encahnements.

The problem though is that if people pull this change and do not realize the new Bool option, two things will happen:

1. They will start getting exceptions/Nothings when reading old Bool data from their tables. 

2. They will write new data to ddb with the new encoding, which will intermix with their old encoding causing a mess that's a bit inconvenient. 

I'll try to make a migration oriented suggestion once I'm in front of a computer. The basic principle here though is that we can't release code that can contaminate database data this suddenly depending on how people were using the library (ie our homebrew boolean encoding).

**Comment by jamesthompson** [comment]
I'm in agreement on all fronts. 

It's worth mentioning, that at present the lack of `BOOL` decoding - i.e. error throwing, is causing exceptions when we pull rows with native dynamo booleans encoded. I feel this is 'bug-like' in it's present implementation given the "in-prod" dynamo at AWS- hence this work.

If the API version is updated with caveats, would that suffice as warnings to library users? I guess the question is - would compile-time errors prevent users from inserting or parsing the old home-rolled boolean values? If that's the case - I think the major API-bump could be justified. Just my two-cents..!

Thanks a lot for being so supportive of this work in general and for the excellent lib!

**Comment by aristidb** [comment]
Shall I unmerge it now, or is it OK to just not make a feature release until we've found a satisfactory solution?

**Comment by ozataman** [comment]
I'm planning to find a way to include an implicit migration of data so that
no existing production setup will start failing or corrupt data in the
transition. I would say if we want to release immediately, we will need to
unmerge. If not, I'll try to squeeze that in before the next release so
unmerging wouldn't be necessary for now.

**Comment by ozataman** [comment]
@jamesthompson Totally agree the current lack of Bool support is bug-like and we definitely need to remedy. We just need to be very careful in this particular case because as we're swapping data representations under people's feet in production databases. This would likely wreak havoc at work for us as an example. 

I know people are supposed to put upper-bounds on their dependencies and look at the release notes (and take necessary action) with each major version bump, but I don't think that happens as ideally as one may think in reality. We should be conservative and take care of our users :)

Here's what I'm thinking:

- We tweak the new Bool parsing instance so that old 0/1 numbers will also be parsed automatically into Bool, but new Bool writes (or over-writes) go through Amazon's new Bool type.
- Need to confirm: Does DynamoDb let us overwrite a previously 0/1 integer-valued key with a True/False value? If DDB does not let us do this, then we will need to rethink this strategy.
- We introduce a newtype around Bool with an instance that keeps the old 0/1 Bool encoding, so people can easily switch to that if they want to keep the representation.
- We need this to coincide with a major bump in version and write prominently in the release notes.

**Comment by aristidb** [comment]
So I did this (not committed yet):

1. Bool instance can parse 0/1
2. New OldBool newtype that writes 0/1, in case anybody wants to stick to the old format
3. I tested it, yes, overwriting a 0/1 with an actual boolean seems to work

So pretty much exactly as @ozataman proposed :)

**Comment by aristidb** [comment]
See 52d7f43

**Comment by jamesthompson** [comment]
Awesome! Thanks everybody for the great work!!

**Comment by ozataman** [comment]
Awesome! Thank you @aristidb for taking care of this!
